lib/src/comment_references/parser.dart

```diff
@@ -0,0 +1,356 @@
+    if (_atEnd) {
+      return _IdentifierParseResult.endOfFile;
+    }
+    var startIndex = _index;
+    while (!_atEnd) {
+      if (_nonIdentifierChars.contains(_thisChar)) {
+        if (startIndex == _index) {
+          return _IdentifierParseResult.notIdentifier;
+        } else {
+          break;
+        }
+      }
+      _index++;
+    }
+    return _IdentifierParseResult.ok(
+        IdentifierNode(codeRef.substring(startIndex, _index)));
+  }
+
+  static const _constructorHintSuffix = '()';
+
+  /// ```text
+  /// <suffix> ::= <constructorPostfixHint>
+  ///   | <trailingJunk>
+  ///
+  /// <trailingJunk> ::= '<'<CHARACTER>*'>'
+  ///   | '('<notClosedParenthesis>*')'
+  ///   | '<'<notGreaterThan>*'>'
+  ///   | '?'
+  ///   | '!'
+  ///
+  /// <constructorPostfixHint> ::= '()'
+  /// ```
+  _SuffixParseResult _parseSuffix() {
+    var startIndex = _index;
+    _walkPastWhitespace();
+    if (_atEnd) {
+      return _SuffixParseResult.missing;
+    }
+    if (_tryMatchLiteral(_constructorHintSuffix)) {
+      if (_atEnd) {
+        return _SuffixParseResult.ok(
+            ConstructorHintEndNode(codeRef.substring(startIndex, _index)));
+      }
+      return _SuffixParseResult.notSuffix;
+    }
+
+    if ((_thisChar == $exclamation || _thisChar == $question) && _nextAtEnd) {
+      return _SuffixParseResult.junk;
+    }
+    if (_matchBraces($lparen, $rparen) || _matchBraces($lt, $gt)) {
+      return _SuffixParseResult.junk;
+    }
+
+    // TODO(jcollins-g): implement junk suffixes if necessary
+    return _SuffixParseResult.notSuffix;
+  }
+
+  bool get _atEnd => _index >= _referenceLength;
+  bool get _nextAtEnd => _index + 1 >= _referenceLength;
+  int get _thisChar => codeRef.codeUnitAt(_index);
+
+  /// Advances [_index] on match, preserves on non-match.
+  bool _tryMatchLiteral(String characters,
+      {bool acceptTrailingWhitespace = true}) {
+    assert(acceptTrailingWhitespace != null);
+    if (characters.length + _index > _referenceLength) return false;
+    for (var startIndex = _index;
+        _index - startIndex < characters.length;
+        _index++) {
+      if (codeRef.codeUnitAt(_index) !=
+          characters.codeUnitAt(_index - startIndex)) {
+        _index = startIndex;
+        return false;
+      }
+    }
+    if (acceptTrailingWhitespace) _walkPastWhitespace();
+    return true;
+  }
+
+  /// Walks past any whitespace characters.
+  ///
+  /// [_index] points to a non-whitespace character when this method returns.
+  ///
+  /// If EOF is reached, then [_index] points past the end of the template
+  /// content when this method returns.
+  void _walkPastWhitespace() {
+    while (!_atEnd) {
+      if (_whitespace.contains(_thisChar)) {
+        _index++;
+      } else {
+        return;
+      }
+    }
+    return;
+  }
+
+  /// Returns true if we started with startChar and ended with endChar
+  /// with a matching number of braces.
+  /// Returns _index to start if returning false.
```

Comment:
"startIndex" would be more meaningful than "start" here.

Reply:
rephrased